bundle/src/main/java/com/adobe/acs/commons/http/headers/impl/ResourceTypeBasedDispatcherMaxAgeHeaderFilter.java

```diff
@@ -0,0 +1,124 @@
+ * limitations under the License.
+ * #L%
+ */
+package com.adobe.acs.commons.http.headers.impl;
+
+import com.day.cq.commons.PathInfo;
+import org.apache.felix.scr.annotations.Component;
+import org.apache.felix.scr.annotations.ConfigurationPolicy;
+import org.apache.felix.scr.annotations.Properties;
+import org.apache.felix.scr.annotations.Property;
+import org.apache.felix.scr.annotations.Reference;
+import org.apache.jackrabbit.JcrConstants;
+import org.apache.sling.api.resource.LoginException;
+import org.apache.sling.api.resource.Resource;
+import org.apache.sling.api.resource.ResourceResolver;
+import org.apache.sling.api.resource.ResourceResolverFactory;
+import org.apache.sling.commons.osgi.PropertiesUtil;
+import org.osgi.service.cm.ConfigurationException;
+import org.osgi.service.component.ComponentContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.servlet.http.HttpServletRequest;
+import java.util.Collections;
+import java.util.Dictionary;
+import java.util.Map;
+
+@Component(
+        label = "ACS AEM Commons - Dispacher Cache Control Header Resource Type Based - Max Age",
+        description = "Adds a Cache-Control max-age header to content based on resource type to enable Dispatcher TTL support.",
+        metatype = true,
+        configurationFactory = true,
+        policy = ConfigurationPolicy.REQUIRE)
+@Properties({
+        @Property(
+                name = "webconsole.configurationFactory.nameHint",
+                value = "Max Age: {max.age} for Resource Types: [{resource.types}]",
+                propertyPrivate = true),
+        @Property(name = "sling.filter.scope", value = {"REQUEST", "FORWARD"})
+})
+public class ResourceTypeBasedDispatcherMaxAgeHeaderFilter extends DispatcherMaxAgeHeaderFilter {
+
+    private static final Logger log = LoggerFactory.getLogger(ResourceTypeBasedDispatcherMaxAgeHeaderFilter.class);
+
+    @Property(label = "Resource types",
+            description = "Resource types the page should have to use this filter.",
+            cardinality = Integer.MAX_VALUE)
+    public static final String PROP_RESOURCE_TYPES = "resource.types";
+
+    private String[] resourceTypes;
+
+    private static final String SERVICE_NAME = "resource-type-dispatcher-header-filter";
+    private static final Map<String, Object> AUTH_INFO;
+
+    static {
+        AUTH_INFO = Collections.singletonMap(ResourceResolverFactory.SUBSERVICE, (Object) SERVICE_NAME);
+    }
+
+    @Reference
+    private ResourceResolverFactory resourceResolverFactory;
+
+    @Override
+    @SuppressWarnings("unchecked")
+    protected boolean accepts(final HttpServletRequest request) {
+        if (!super.accepts(request)) {
+            log.debug("Not accepting request because it is not coming from the dispatcher.");
+            return false;
+        }
+
+        try (ResourceResolver resourceResolver = resourceResolverFactory.getServiceResourceResolver(AUTH_INFO)) {
```

Comment:
Is a service user required here? Doesn't the request require enough security context to be able to derive the Resource type anyhow?

Could this simply be:

```
return resourceTypes.stream().filter(rt -> { 
    if(slingRequest.getResource().isResourceType(rt)) { 
       return true; 
   } else if (slingRequest.getResource().getChild("jcr:content") != null) {
       return slingRequest.getResource().getChild("jcr:content").isResourceType(rt);
   } ).anyMatch(); 
``

Reply:
It would have been that simple, if the dispatcher filters were sling filters, but these are registered through the http whiteboard so they don't have a SlingHttpServletRequest but a normal HttpServletRequest

Reply:
oh they are? the `@Property(name = "sling.filter.scope", value = {"REQUEST", "FORWARD"})` threw me ... i assumed they were Sling Filters .. but it does appear theyre registered with the OSGi HTTP Whiteboard...  

@bstopp  you remember why these aren't Sling? is it worth/could we safely switch these over to register as Sling Filters as it removes the need for a service user ..

Reply:
Converting to Sling Filters would make sense here. The notion of resource types only exists in Sling, and not on fetures outside of Sling (CRX/DE, Package Manager, OSGI Webconsole).